zerver/tests/test_narrow.py

```diff
@@ -504,6 +519,36 @@ def dr_emails(dr):
             for message in result["messages"]:
                 self.assertEqual(dr_emails(message['display_recipient']), emails)
 
+    def test_get_messages_with_narrow_huddle_with(self):
+        # type: () -> None
+        """
+        A request for old messages with a narrow by huddle-with only returns
+        group-private conversations with that user.
+        """
+        me = 'hamlet@zulip.com'
+
+        def dr_emails(dr):
+            # type: (Union[Text, List[Dict[str, Any]]]) -> List[Any]
+            assert isinstance(dr, list)
+            emails = []
+            for r in dr:
+                emails.append(r['email'])
+            return emails
+
+        self.send_message(me, ['iago@zulip.com', 'cordelia@zulip.com'], Recipient.HUDDLE)
+
+        personals = [m for m in get_user_messages(get_user_profile_by_email(me))
+                     if m.recipient.type == Recipient.HUDDLE]
+
+        for personal in personals:
+            self.login(me)
+            emails = dr_emails(get_display_recipient((personal.recipient)))
+            for email in emails:
+                narrow = [dict(operator='huddle-with', operand=email)]
+                results = self.get_and_check_messages(dict(narrow=ujson.dumps(narrow)))
+                for message in results["messages"]:
+                    self.assertIn(email, dr_emails(message['display_recipient']))
```

Comment:
I'd like to make the testing here a bit more extensive and systematic to make sure we cover all the corner cases.  Ideally, we should check these conditions (assuming we're hamlet and searching for `huddle-with:cordelia@zulip.com`):
* PM messages between hamlet+cordelia are not included
* Huddle message including cordelia+othello+iago are not included
* Huddle message including hamlet+othello+iago are not included
* Huddle message including cordelia+othello+iago are not included
* Huddle message including hamlet+cordelia+othello are included
* Huddle message including hamlet+cordelia+othello+iago are included

If we do that, we can be pretty confident that this logic works in all cases (we'll likely want a similar approach in the frontend testing).

So I'd build an appropriate loop structure for this test that can check all 5 of those test cases without a ton of test code.